client/state-db/src/pruning.rs

```diff
@@ -24,418 +24,1031 @@
+			// assume that parent was canonicalized
+			self.base = number;
+		} else if (self.base + self.queue.len() as u64) != number {
+			return Err(Error::StateDb(StateDbError::InvalidBlockNumber))
+		}
 		trace!(target: "state-db", "Adding to pruning window: {:?} ({} inserted, {} deleted)", hash, commit.data.inserted.len(), commit.data.deleted.len());
-		let inserted = if self.count_insertions {
+		let inserted = if matches!(self.queue, DeathRowQueue::Mem { .. }) {
 			commit.data.inserted.iter().map(|(k, _)| k.clone()).collect()
 		} else {
 			Default::default()
 		};
 		let deleted = ::std::mem::take(&mut commit.data.deleted);
 		let journal_record = JournalRecord { hash: hash.clone(), inserted, deleted };
-		let block = self.pending_number + self.death_rows.len() as u64;
-		let journal_key = to_journal_key(block);
-		commit.meta.inserted.push((journal_key.clone(), journal_record.encode()));
-		self.import(
-			&journal_record.hash,
-			journal_key,
-			journal_record.inserted.into_iter(),
-			journal_record.deleted,
-		);
+		commit.meta.inserted.push((to_journal_key(number), journal_record.encode()));
+		self.queue.import(self.base, journal_record);
 		self.pending_canonicalizations += 1;
+		Ok(())
 	}
 
 	/// Apply all pending changes
 	pub fn apply_pending(&mut self) {
 		self.pending_canonicalizations = 0;
 		for _ in 0..self.pending_prunings {
 			let pruned = self
-				.death_rows
-				.pop_front()
-				.expect("pending_prunings is always < death_rows.len()");
+				.queue
+				.pop_front(self.base)
+				// NOTE: `pop_front` should not return `MetaDb::Error` because blocks are visited
+				// by `RefWindow::prune_one` first then `RefWindow::apply_pending` and
+				// `DeathRowQueue::get` should load the blocks into cache already
+				.expect("block must loaded in cache thus no MetaDb::Error")
+				.expect("pending_prunings is always < queue.len()");
 			trace!(target: "state-db", "Applying pruning {:?} ({} deleted)", pruned.hash, pruned.deleted.len());
-			if self.count_insertions {
-				for k in pruned.deleted.iter() {
-					self.death_index.remove(k);
-				}
-			}
-			self.pending_number += 1;
+			self.base += 1;
 		}
 		self.pending_prunings = 0;
 	}
 
 	/// Revert all pending changes
 	pub fn revert_pending(&mut self) {
-		// Revert pending deletions.
-		// Note that pending insertions might cause some existing deletions to be removed from
-		// `death_index` We don't bother to track and revert that for now. This means that a few
-		// nodes might end up no being deleted in case transaction fails and `revert_pending` is
-		// called.
-		self.death_rows.truncate(self.death_rows.len() - self.pending_canonicalizations);
-		if self.count_insertions {
-			let new_max_block = self.death_rows.len() as u64 + self.pending_number;
-			self.death_index.retain(|_, block| *block < new_max_block);
-		}
+		self.queue.revert_recent_add(self.base, self.pending_canonicalizations);
 		self.pending_canonicalizations = 0;
 		self.pending_prunings = 0;
```

Comment:
@arkpar not exactly sure why we change pending_pruning here (was in the original code).

Reply:
I thought the reason may be the `CommitSet` that had the data collected in `prune_one` is fail to commit to the db, reset the `pending_prunings` could allow us to re-collect the data in `prune_one` and retry.